### ISSUE TITLE
PyPy 3.8 and 3.9 are sporadically crashing when running test suite with pytest-xdist (but not PyPy 3.7 or any CPython version)

### ISSUE THREAD

**Issue by gitlab-importer** [pypy/pypy#3903]
In Heptapod by @jab on Mar 13, 2023, 23:31

Reporting this here in the hope that it's helpful.

I've been running [bidict's test suite](https://github.com/jab/bidict/tree/main/tests) against [PyPy 3.7, 3.8, and 3.9 (via GitHub Actions' setup-python action)](https://github.com/jab/bidict/blob/141446ec2054bbe16d7dadeb1c0ed69cbc82aed8/.github/workflows/test.yml#L35-L37) over the last year+. (And for the last few months, using pytest-xdist with `--num-processes=auto` to speed up runs.) Over the past couple weeks, the nightly PyPy 3.8 and 3.9 runs have occasionally crashed right as they were finishing up an otherwise successful run, but then the same job eventually succeeded after 1-3 retries.

Here are some examples:
* https://github.com/jab/bidict/actions/runs/4270832569/jobs/7434886888#step:8:380
* https://github.com/jab/bidict/actions/runs/4234711135/jobs/7364288283#step:8:368
* https://github.com/jab/bidict/actions/runs/4234711135/jobs/7364288283#step:8:363
* https://github.com/jab/bidict/actions/runs/4398364046/jobs/7702162876#step:8:351
* https://github.com/jab/bidict/actions/runs/4398364046/jobs/7727569095#step:8:351

pytest-xdist worker number varied between 0 and 1, and PyPy version varied between 3.8 and 3.9 (interestingly, PyPy 3.7 has always succeeded).

The complete error message was the same every time:
```
Fatal Python error: Aborted

Stack (most recent call first, approximate line numbers):
  File "/home/runner/work/bidict/bidict/.tox/py/lib/pypy3.9/site-packages/hypothesis/vendor/pretty.py", line 636 in _repr_pprint
tests/property_tests/test_properties.py::test_views 
[gw0] node down: Not properly terminated
[gw0] [100%] FAILED tests/property_tests/test_properties.py::test_views
```

(All tests always succeeded up to this point, and `tests/property_tests/test_properties.py::test_views` is always the last test to run given consistent ordering of the collected tests.)

I understand that `Fatal Python error: Aborted` means that the Python interpreter crashed while executing the test, and usually you have to use a ctypes/cffi/C extension to cause that. Since bidict is implemented in pure Python, PyPy 3.8 and 3.9 have been crashing but all other interpreters running the tests aren't, I thought I'd report this here in case it's a bug in PyPy. Please let me know if there's any other info I can provide, and thanks for all the great work on this amazing project. :heart:

**Comment by jab** [comment]
Haven't seen this in a while, though I'm not sure what may have fixed it. Please feel free to close in any case. I can report back here if I ever see it again.

**Comment by jab** [comment]
I just came across https://www.pypy.org/posts/2024/03/fixing-bug-incremental-gc.html and am wondering if that might have been the cause of this issue?

Coincidentally, I just saw [PyPy 3.9 segfault in CI](https://github.com/jab/bidict/actions/runs/8453317922/job/23155725496#step:11:13) for the first time in ages:
```
  env:
    pythonLocation: /opt/hostedtoolcache/PyPy/3.9.18/x64
    ...
/home/runner/work/_temp/d2aa6e83-fa7c-4822-9533-174f8ed9eac3.sh: line 1:  1863 Segmentation fault      (core dumped) ${RUN_PYTEST_CMD:-python} -m pytest
Error: Process completed with exit code 139.
```
(I changed how the tests are invoked in CI from using `tox` to the above since I created this issue, but not sure how that affects the manifestation of any GC bugs.)

P.S. Thanks for the heroic bug-fixing and great writeup, @cfbolz, and thanks to the entire PyPy team!
